python/common/org/python/stdlib/datetime/DateTime.java

```diff
@@ -0,0 +1,377 @@
+package org.python.stdlib.datetime;
+
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Collections;
+
+import org.python.types.Int;
+
+import java.util.Calendar;
+import python.datetime;
+
+public class DateTime extends org.python.types.Object {
+	private final int YEAR_INDEX = 0;
+	private final int MONTH_INDEX = 1;
+	private final int DAY_INDEX = 2;
+	private final int HOUR_INDEX = 3;
+	private final int MINUTE_INDEX = 4;
+	private final int SECOND_INDEX = 5;
+	private final int MICROSECOND_INDEX = 6;
+
+	private final int MIN_YEAR = 1;
+	private final int MAX_YEAR = 9999;
+
+	private long[] timeUnits = { 0l, 0l, 0l, 0l, 0l, 0l, 0l };
+
+	@org.python.Attribute
+	public final org.python.Object year;
+
+	@org.python.Attribute
+	public final org.python.Object month;
+
+	@org.python.Attribute
+	public final org.python.Object day;
+
+	@org.python.Attribute
+	public final org.python.Object hour;
+
+	@org.python.Attribute
+	public final org.python.Object minute;
+
+	@org.python.Attribute
+	public final org.python.Object second;
+
+	@org.python.Attribute
+	public final org.python.Object microsecond;
+
+	@org.python.Attribute
+	public static final org.python.Object min = __min__();
+
+	@org.python.Attribute
+	public static final org.python.Object max = __max__();
+
+	public DateTime(org.python.Object[] args, java.util.Map<java.lang.String, org.python.Object> kwargs) {
+		super();
+		String[] keys = { "year", "month", "day", "hour", "minute", "second", "microsecond" };
+		boolean kwargsIsUsed = false;
+		int keyIndex = 0;
+		int argIndex = 0;
+
+		for (String key : keys) {
+			if (kwargs.get(key) != null) {
+				this.timeUnits[keyIndex] = ((org.python.types.Int) kwargs.get(key)).value;
+				kwargsIsUsed = true;
+			} else if (args.length > argIndex) {
+				if (kwargsIsUsed)
+					throw new org.python.exceptions.SyntaxError("positional argument follows keyword argument");
+				this.timeUnits[keyIndex] = ((org.python.types.Int) args[argIndex]).value;
+				argIndex++;
+			} else if (keyIndex < 3) {
+				throw new org.python.exceptions.TypeError(
+						"Required argument '" + keys[keyIndex] + "' (pos " + (keyIndex + 1) + ") not found");
+			}
+			keyIndex++;
+		}
+
+		if (this.timeUnits[YEAR_INDEX] < MIN_YEAR || this.timeUnits[YEAR_INDEX] > MAX_YEAR) {
+			throw new org.python.exceptions.ValueError("year " + this.timeUnits[YEAR_INDEX] + "is out of range");
+		}
+
+		if (this.timeUnits[MONTH_INDEX] < 1 || this.timeUnits[MONTH_INDEX] > 12) {
+			throw new org.python.exceptions.ValueError("month " + this.timeUnits[MONTH_INDEX] + "is out of range");
+		}
+		if (this.timeUnits[DAY_INDEX] < 1 || this.timeUnits[DAY_INDEX] > 31) {
```

Comment:
This will yield unexpected behaviour for months with less than 31 days. Also, with this february can have more than 28 days

Reply:
Unexpected behaviour a.k.a invalid dates :D

Reply:
true, fixed now